src/main/webapp/app/overview/course-card.component.ts

```diff
@@ -60,7 +60,14 @@ export class CourseCardComponent implements OnChanges {
     ngOnChanges() {
         if (this.course.exercises && this.course.exercises.length > 0) {
             this.exerciseCount = this.course.exercises.length;
-            const nextExercises = this.exerciseService.getNextExercisesForDays(this.course.exercises);
+            // sets the exercise with the earliest due date which is not yet successful (no 100%)
+
+            // debugger;
+            const nextExercises = this.exerciseService.getNextExercisesForDays(
+                this.course.exercises!.filter((exercise: Exercise) => !exercise.studentParticipations?.[0]?.submissions?.[0]?.results?.[0]?.successful),
```

Comment:
Are you sure that this lines works? What happens if a student participated in an exercise but didn't generate a submission? I might be wrong but I should you make sure that the arrays are accessible before accessing them with an index?

Reply:
You mean, what happens if the student just opened an exercise (with/without writing anything) and does not click the "submit" button? 
I tried it out and it still worked.
I am (very) new to (JS, TS, etc.) but if I am not mistaken, e.g. "student?" returns a student or undefined (student? -> student | undefined). 
My goal is to remove every exercise which is successful. So everything else will **not** be removed, i.e. undefined as well. So the use case you mentioned should still be displayed, because it will not be removed.
e.g. no submissions -> submission array will be empty : 
=> !exercise.studentParticipations?.[0]?.submissions?.[0]? // submissions exists but is empty -> submissions[0] = undefined
(If I am not mistaken) undefined == false -> the "!" in front will turn undefined into true and not remove/filter it out 
If you disagree feel free to educate me on the topic. 🙂

Reply:
It looks to me like this line should work, since I think the `?.` operator catches the array being undefined before access. But I see another issue: Don't we have to check all results? It could be that the first result is not successful, but the second one (e.g. index 1 then). Same for submissions. I'm also not sure if `studentParticipations` are just the participations for the current student, or if they are for more. (Which wouldn't make much sense for student accounts, but might be the case for tutor or instructor accounts? If so, is this an edge case we need to handle?)

Reply:
Hm, I tried provoking a case where this breaks, but couldn't find one, so maybe this does work as is. (Unfortunately I'm not too familiar with our result handling.)

Reply:
Ah yes the code should indeed work since you use the ? before accessing and after accessing the array elements :)